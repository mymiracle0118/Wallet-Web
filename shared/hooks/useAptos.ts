import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getAptosWalletUsingSeed, getBalance } from '@portal/shared/services/aptosService'
import { ethers } from 'ethers'
import { toUTF8Array } from '../utils/utf8'

type APTOSWalletState = {
  /** APTOS network - create wallet */
  createAptosWallet: (mnemonics?: string) => Promise<any>
  /** APTOS network - Update balance */
  refreshBalance: (aptosType?: string) => Promise<void>
  _keypair?: Ed25519Keypair | null
  /** APTOS network - Wallet Address */
  address?: string
  /** APTOS network - privateKey Address */
  privateKey?: string
  /** APTOS network - Total Balance of wallet */
  balance: number
  /** APTOS network - All objects owned by wallet */
  objectsOwned?: Array<any>

  wallet?: ethers.Wallet
  encryptedPrivateKey: (string | number)[]
  encryptedWallet?: string
}

const initialState = {
  address: undefined,
  balance: 0,
  _keypair: undefined,
  privateKey: undefined,
  encryptedPrivateKey: [],
}

export const useAptos = create<APTOSWalletState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,

        createAptosWallet: async (mnemonics?: string) => {
          const wallet = await getAptosWalletUsingSeed(mnemonics)
          if (wallet) {
            const encryptedPrivateKey = toUTF8Array(wallet?.privateKey)
            set({ address: wallet?.address, encryptedPrivateKey })
          }

          const { refreshBalance } = get()
          await refreshBalance()
          return wallet
        },
        refreshBalance: async (aptosType?: string) => {
          const { address } = get()
          if (!address) return
          const balance = await getBalance(address, aptosType)
          set({ balance: balance })
        },
      }),
      {
        name: 'APTOS-storage',
        getStorage: () => chromeSyncStorage,
        partialize: (state) => Object.fromEntries(Object.entries(state).filter(([key]) => !['wallet'].includes(key))), // Prevents decrypted wallet from being persisted in storage
      }
    ),
    { name: 'APTOS - Wallet' }
  )
)
