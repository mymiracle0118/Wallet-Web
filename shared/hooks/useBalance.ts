import { useEffect, useMemo } from 'react'
import { useQuery } from '@tanstack/react-query'
import { BigNumber } from 'ethers'

import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

type BalanceResponse = {
  balance: BigNumber | number
  balanceFormatted: number
  isLoading: boolean
}
const { updateTokenBalance } = useStore.getState()
export const useBalance = (token: NetworkToken): BalanceResponse => {
  let balance, balanceFormatted

  const networkFactory = NetworkFactory.selectByNetworkId(token.shortName)
  if (!networkFactory) {
    return {
      balance: 0,
      balanceFormatted: 0,
      isLoading: false,
    }
  }

  const walletAddress = token.address
  const { data, isLoading } = useQuery(
    ['asset-balance', token.networkName, walletAddress, token.shortName],
    () => networkFactory.getTokenBalance(walletAddress as string, token),
    {
      refetchInterval: 60000,
      staleTime: 30000,
    }
  )

  balance = useMemo(() => {
    return data || BigNumber.from(0)
  }, [data])
  balanceFormatted = useMemo(() => networkFactory.getBalanceFormattedValue(data), [data]) as number

  useEffect(() => {
    // Update asset balance in wallet to keep running total
    if (balance && balanceFormatted) {
      updateTokenBalance(token.shortName, balance, balanceFormatted)
    }
  }, [balance, balanceFormatted])

  return {
    balance,
    balanceFormatted,
    isLoading,
  }
}
