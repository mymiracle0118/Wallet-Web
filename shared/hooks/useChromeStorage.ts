import { useCallback, useEffect, useState } from 'react'

export const storage = {
  get: (key: string, defaultValue: any, storageArea: Storage) => {
    const keyObj = defaultValue === undefined ? key : { [key]: defaultValue }
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].get(keyObj, (items: any) => {
        const error = chrome.runtime.lastError
        if (error) return reject(error)
        resolve(items[key])
      })
    })
  },
  set: (key: string, value: any, storageArea: Storage) => {
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].set({ [key]: value }, () => {
        const error = chrome.runtime.lastError
        error ? reject(error) : resolve()
      })
    })
  },
}

/**
 * Basic hook for storage
 * @param {string} key
 * @param {*} initialValue
 * @param {'local'|'sync'} storageArea
 * @returns {[*, function(*= any): void, boolean, string]}
 */

export default function useChromeStorage(key, initialValue, storageArea) {
  const [INITIAL_VALUE] = useState(() => {
    return typeof initialValue === 'function' ? initialValue() : initialValue
  })
  const [STORAGE_AREA] = useState(storageArea)
  const [state, setState] = useState(INITIAL_VALUE)
  const [isPersistent, setIsPersistent] = useState(true)
  const [error, setError] = useState('')

  useEffect(() => {
    storage
      .get(key, INITIAL_VALUE, STORAGE_AREA)
      .then((res) => {
        setState(res)
        setIsPersistent(true)
        setError('')
      })
      .catch((error) => {
        setIsPersistent(false)
        setError(error)
      })
  }, [key, INITIAL_VALUE, STORAGE_AREA])

  const updateValue = useCallback(
    (newValue) => {
      const toStore = typeof newValue === 'function' ? newValue(state) : newValue
      storage
        .set(key, toStore, STORAGE_AREA)
        .then(() => {
          setIsPersistent(true)
          setError('')
        })
        .catch((error) => {
          // set newValue to local state because chrome.storage.onChanged won't be fired in error case
          setState(toStore)
          setIsPersistent(false)
          setError(error)
        })
    },
    [STORAGE_AREA, key, state]
  )

  useEffect(() => {
    const onChange = (changes, areaName) => {
      if (areaName === STORAGE_AREA && key in changes) {
        setState(changes[key].newValue)
        setIsPersistent(true)
        setError('')
      }
    }
    chrome.storage.onChanged.addListener(onChange)
    return () => {
      chrome.storage.onChanged.removeListener(onChange)
    }
  }, [key, STORAGE_AREA])

  return [state, updateValue, isPersistent, error]
}
