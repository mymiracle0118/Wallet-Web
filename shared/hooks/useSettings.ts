import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import produce from 'immer'

import type { Asset } from './useWallet'
import { AccountWallet, useWallet } from './useWallet'
import { SAFETY_MEASURE } from '@portal/portal-extension/src/utils/constants'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { setLockUp } from './useLockUp'
import { EnvironmentType, NetworkToken } from '@portal/shared/utils/types'
import { useStore } from './useStore'

export type AddressBookUser = {
  avatar?: string
  username: string
  address: string
  safety?: SAFETY_MEASURE
  network: string
}
export type AccountTokenType = {
  isFavorite: boolean
}
export type TAccount = {
  assets: Array<Asset>
  address: string
  username: string
  encryptedPrivateKey: string
  encryptedWallet: string
  avatar: string
  balance: number
  isAccountImported: boolean
  networkName: string
  tokens: { [env: string]: { [key: string]: AccountTokenType } }
  customTokens: { [env: string]: { [key: string]: NetworkToken } }
  isPrimary: boolean
}

type TAddressBook = {
  username: string
  address: string
}

export type SettingsState = {
  enablePasswordProtection: boolean
  accounts: { [address: string]: TAccount }
  currentAccount: TAccount | null
  darkMode: boolean
  addressBook: AddressBookUser[]
  saveAccount: (
    username: string,
    address: string,
    networkName: string,
    isPrimary?: boolean,
    isAccountImported?: boolean
  ) => void
  clearAccounts: () => void
  clearAddressbook: () => void
  setDarkMode: (mode: boolean) => void
  setEnablePasswordProtection: (enable: boolean) => void
  hideAccount: (address: string) => void
  addToAddressBook: (user: AddressBookUser) => void
  removeToAddressBook: (address: string) => void
  updateUsername: (username: string, address: string) => void
  networkEnvironment: EnvironmentType
  setNetworkEnvironment: (envType: EnvironmentType) => void
  setCurrentAccount: (account: TAccount) => void
  addToken: (accountKey: string, tokenKey: string) => void
  updateTokens: (accountKey: string, tokenKeys: string[]) => void
  addRemoveFavoriteToken: (accountKey: string, tokenKey: string, isFavoriteValue: boolean) => void
  addCustomToken: (token: NetworkToken) => void
}

export const useSettings = create<SettingsState>()(
  devtools(
    persist(
      (set, get) => ({
        accounts: {},
        currentAccount: null,
        addressBook: [],
        enablePasswordProtection: false,
        darkMode: true,
        networkEnvironment: 'testNet',
        setDarkMode: (mode) => {
          set({ darkMode: mode })
        },

        updateUsername: async (newUsername: string, address: string) => {
          set(
            produce((state: SettingsState) => {
              state.accounts[address].username = newUsername
            })
          )
        },

        addToAddressBook: (user: AddressBookUser) => {
          set(
            produce((state: SettingsState) => ({
              addressBook: [...state.addressBook, user],
            }))
          )
        },

        removeToAddressBook: (address: string) => {
          set(
            produce((state: SettingsState) => {
              state.addressBook = state.addressBook.filter((v) => v.address !== address)
            })
          )
        },

        saveAccount: async (username, address, networkName, isPrimary, isAccountImported) => {
          const { avatar, encryptedWallet } = useWallet.getState()
          const { addTokenToList } = useStore.getState()
          if (username && address) {
            const account = {
              address,
              username,
              encryptedWallet,
              avatar,
              networkName,
              isAccountImported: isAccountImported ? isAccountImported : false,
              tokens: { testNet: {}, mainNet: {} },
              customTokens: { testNet: {}, mainNet: {} },
              isPrimary: isPrimary || false,
            }
            set(
              produce((state: SettingsState) => {
                // @ts-ignore
                state.accounts[address] = account
              })
            )
            const { accounts, currentAccount } = get()
            if (!currentAccount || (currentAccount && !currentAccount.address)) {
              set({ currentAccount: accounts[address] })
            }
            addTokenToList(address, networkName)
          }
        },

        clearAccounts: () => {
          set({ accounts: {}, currentAccount: null })
        },

        clearAddressbook: () => {
          set({ addressBook: [] })
        },

        setEnablePasswordProtection: (enable: boolean) => {
          setLockUp(enable)
          set({
            enablePasswordProtection: enable,
          })
        },

        hideAccount: (addressHide: string) => {
          const accountLists = { ...get().accounts }
          if (accountLists[`${addressHide}`]) {
            delete accountLists[`${addressHide}`]
          }

          set(
            produce((state: SettingsState) => {
              state.accounts = accountLists
            })
          )

          const { walletsList } = useStore.getState()
          if (walletsList[`${addressHide}`]) {
            delete walletsList[`${addressHide}`]
            useStore.getState().removeWalletFromWalletList(walletsList)
          }

          const { address } = useWallet.getState()
          if (address) {
            const account = useSettings.getState().accounts[address]
            useSettings.getState().setCurrentAccount(account)
          }
        },
        setNetworkEnvironment: (envType: EnvironmentType) => {
          set(
            produce((state: SettingsState) => {
              state.networkEnvironment = envType
            })
          )
        },
        setCurrentAccount: (account: TAccount) => {
          const { updateCurrentAccountTokensList } = useStore.getState()
          set(
            produce((state: SettingsState) => {
              state.currentAccount = account
            })
          )
          updateCurrentAccountTokensList(account.address)
        },
        addToken: (accountKey, tokenKey: string) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingToken = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              const newToken: AccountTokenType = { isFavorite: false }
              existingToken[tokenKey] = newToken
              state.accounts[accountKey].tokens[networkEnvironment] = existingToken
            })
          )
        },
        addRemoveFavoriteToken: (accountKey: string, tokenKey: string, isFavoriteValue: boolean) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingToken = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              const newToken: AccountTokenType = { isFavorite: false }
              existingToken[tokenKey] = { ...existingToken[tokenKey], isFavorite: isFavoriteValue }
              state.accounts[accountKey].tokens[networkEnvironment] = existingToken
            })
          )
        },
        updateTokens: (accountKey: string, tokenKeys: string[]) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingTokens = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              for (const key in existingTokens) {
                if (!tokenKeys.includes(key)) {
                  delete existingTokens[key]
                }
              }
              state.accounts[accountKey].tokens[networkEnvironment] = existingTokens
            })
          )
        },
        addCustomToken: async (token: NetworkToken) => {
          const { networkEnvironment, currentAccount } = get()
          if (currentAccount) {
            const accountKey = currentAccount.address
            set(
              produce((state: SettingsState) => {
                if (!state.accounts[accountKey].customTokens[networkEnvironment]) {
                  state.accounts[accountKey].customTokens[networkEnvironment] = {}
                }
                const existingToken = { ...state.accounts[accountKey].customTokens[networkEnvironment] }
                existingToken[token.shortName] = token
                state.accounts[accountKey].customTokens[networkEnvironment] = existingToken
              })
            )

            const { addTokenToList, addWalletToList, walletsList } = useStore.getState()

            const newWalletObj: AccountWallet = { ...walletsList[accountKey]['ETH'] }
            await addWalletToList(accountKey, token.networkName, newWalletObj)
            addTokenToList(accountKey, token.shortName)
          }
        },
      }),
      {
        name: 'account-storage',
        getStorage: () => chromeSyncStorage,
      }
    ),
    { name: 'Shuttle - Accounts' }
  )
)
