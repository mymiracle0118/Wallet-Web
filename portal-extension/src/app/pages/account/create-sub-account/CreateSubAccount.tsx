import { useSettings } from '@portal/shared/hooks/useSettings'
import { useWallet } from '@portal/shared/hooks/useWallet'
import { Button, CustomTypography, Form, Input } from 'components'
import SinglePageLayout from 'layouts/single-page-layout/SinglePageLayout'
import { goBack } from 'lib/woozie'
import { useState } from 'react'
import { useTranslation } from 'react-i18next'

import { yupResolver } from '@hookform/resolvers/yup'
import { Avatar } from '@nextui-org/react'
import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { useStore } from '@portal/shared/hooks/useStore'
import ChangeProfileAvatar from '@src/app/components/ChangeProfileAvatar'
import { SpinnerIcon } from '@src/app/components/Icons'
import PasswordPromptModal from 'app/pages/wallet/PasswordPromptModal'
import defaultAvatar from 'assets/images/Avatar.png'
import { ethers } from 'ethers'
import { useForm } from 'react-hook-form'
import * as yup from 'yup'

const CreateSubAccount = () => {
  const { t } = useTranslation()
  const [loading, setLoading] = useState<boolean>(false)
  const { storeWallet, changeAvatar } = useWallet()
  const { saveAccount, accounts } = useSettings()
  const { walletsList } = useStore()
  const [isShowAvatarModal, setShowAvatarModal] = useState<boolean>(false)
  const [selectedAvatar, setSelectedAvatar] = useState<null | string>(null)
  const [openPasswordModal, setOpenPasswordModal] = useState<boolean>(false)
  const [subUsername, setSubUsername] = useState('')

  const schema = yup.object().shape({
    username: yup
      .string()
      .min(3, t('Account.usernameMinimum'))
      .max(15, t('Account.usernameMaximum'))
      .required(t('Account.usernameRequired'))
      .matches(/^\S*$/, t('Account.usernameNoSpace')),
  })
  const methods = useForm({
    resolver: yupResolver(schema),
    mode: 'onChange',
  })
  const {
    register,
    handleSubmit,
    setError,
    watch,
    formState: { errors, isValid, isDirty },
  } = methods

  const handleCreateAccount = ({ username }: { username: string }) => {
    if (selectedAvatar) {
      changeAvatar(selectedAvatar)
    }
    setSubUsername(username)
    setOpenPasswordModal(true)
  }

  const handlePasswordPromptSuccess = async (password: string) => {
    setOpenPasswordModal(false)
    try {
      setLoading(true)
      const primaryAccount = Object.values(accounts).find((acc) => acc.isPrimary)
      if (primaryAccount) {
        const accountWallet = walletsList[primaryAccount.address][primaryAccount.networkName]
        const currentWallet = await ethers.Wallet.fromEncryptedJson(
          accountWallet.encryptedWallet as string,
          password as string
        )
        if (currentWallet) {
          const { wallet, encryptedPrivateKey, deriveIndex } = await NetworkFactory.checkAndCreateNextDeriveWallet(
            currentWallet.mnemonic.phrase as string,
            primaryAccount.networkName
          )

          await storeWallet(
            { wallet, encryptedPrivateKey, deriveIndex },
            subUsername,
            password,
            primaryAccount.networkName,
            wallet.address
          )
          saveAccount(subUsername, wallet.address, primaryAccount.networkName)
        }
      }
      setLoading(false)
      goBack()
    } catch (error) {
      setLoading(false)
      let message = 'Something is wrong!'
      if (error instanceof Error) message = error.message
      setError('username', { type: 'custom', message: message })
    }
  }

  return (
    <SinglePageLayout title="Create sub account">
      <Form methods={methods} onSubmit={handleSubmit(handleCreateAccount)}>
        <div className="space-y-4">
          <div className="flex flex-col items-center space-y-4">
            <Avatar src={selectedAvatar || defaultAvatar} className="w-24 h-24 rounded-full" />

            <Button
              variant="bordered"
              color="outlined"
              className="font-extrabold h-11 w-24"
              onClick={() => setShowAvatarModal(true)}
            >
              {t('Actions.change')}
            </Button>
          </div>

          <div className="space-y-1">
            <CustomTypography variant="subtitle" type="secondary">
              {t('Onboarding.accountName')}
            </CustomTypography>
            <Input
              autoComplete="off"
              dataTestId="username"
              mainColor
              id="username"
              placeholder={t('Onboarding.accountName')}
              endAdornment={
                <div
                  className={`text-[12px] ${
                    watch('username')?.length > 15 ? 'text-feedback-negative' : 'text-fotter-dark-inactive'
                  }`}
                >
                  {watch('username')?.length || 0}/15
                </div>
              }
              fullWidth
              disabled={loading}
              {...register('username')}
              error={errors.username?.message}
              dynamicColor={'text-feedback-negative'}
            />
          </div>
        </div>

        <div className="flex gap-2 justify-between mt-8">
          <Button variant="bordered" color="outlined" onClick={() => goBack()}>
            {t('Actions.cancel')}
          </Button>
          <Button
            type="submit"
            data-test-id="create-subaccount-btn"
            color={!isDirty || !isValid || loading || subUsername?.length > 4 ? 'disabled' : 'primary'}
            isDisabled={!isDirty || !isValid || loading || subUsername?.length > 4}
            isLoading={loading}
            spinner={<SpinnerIcon />}
          >
            {!loading && t('Actions.create')}
          </Button>
        </div>

        <ChangeProfileAvatar
          openModal={isShowAvatarModal}
          closeModal={() => {
            setShowAvatarModal(false)
            setSelectedAvatar('')
          }}
          selectedAvatar={selectedAvatar}
          setSelectedAvatar={setSelectedAvatar}
          handleAvatarChange={() => setShowAvatarModal(false)}
        />
      </Form>
      <PasswordPromptModal
        modalState={openPasswordModal}
        closePromptModal={() => setOpenPasswordModal(false)}
        onPromptPassword={handlePasswordPromptSuccess}
      />
    </SinglePageLayout>
  )
}

export default CreateSubAccount
