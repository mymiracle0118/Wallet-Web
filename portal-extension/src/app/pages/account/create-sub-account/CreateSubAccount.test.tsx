/* eslint-disable @typescript-eslint/no-unsafe-call */

import React from 'react'
import userEvent from '@testing-library/user-event'
import { render, screen, renderHook, act } from '@testing-library/react'
import { useWallet } from '@portal/shared/hooks/useWallet'
import { useSettings } from '@portal/shared/hooks/useSettings'
import CreateSubAccount from 'pages/account/create-sub-account/CreateSubAccount'
import { I18nextProvider } from 'react-i18next'
import i18n from '@portal/shared/i18n/config'

describe('<CreateSubAccount />', () => {
  test('renders correctly Create Sub Account Page', () => {
    const tree = render(<CreateSubAccount />)
    expect(tree).toMatchSnapshot()
  })

  test('Create Sub Account functionality', async () => {
    const user = userEvent.setup()
    const { result: walletResult } = renderHook(() => useWallet())
    const { result: walletSettings } = renderHook(() => useSettings())

    act(() => {
      walletResult.current.address = '0x6886e372363bA8718993EbF49f1a7324F9b273bd'
      walletResult.current.encryptedWallet =
        '{"address":"6886e372363ba8718993ebf49f1a7324f9b273bd","id":"62d90545-71b7-4ed3-9338-6cb2b587f475","version":3,"Crypto":{"cipher":"aes-128-ctr","cipherparams":{"iv":"b5f0c3d1d24cd9cea36528de0605bad3"},"ciphertext":"887a2a44586a339a347379f3107580cb8c2709d832b1628d499a0cb4ad2223ef","kdf":"scrypt","kdfparams":{"salt":"c3d6770af197cecc644086460bf1f9ddb504cd993eb903e49d2ada7714b5b381","n":65536,"dklen":32,"p":1,"r":8},"mac":"b2517acaa5781bb513a0087b24cb933d7228641bb12a716fa8df50091773fb11"},"x-ethers":{"client":"ethers.js","gethFilename":"UTC--2022-07-01T06-29-13.0Z--6886e372363ba8718993ebf49f1a7324f9b273bd","mnemonicCounter":"fb5ea5b9bc34a33a0ca6deb287333f0c","mnemonicCiphertext":"35bffda94500f27dfc6245f0c3ba7cc0","path":"m/44\'/60\'/0\'/0/0","locale":"en","version":"0.1"}}'
      walletResult.current.lockTime = 5
      walletResult.current.lockWallet = false
      walletResult.current.username = 'Test 1'

      walletSettings.current.accounts = [
        {
          address: '0x6886e372363bA8718993EbF49f1a7324F9b273bd',
          encryptedWallet:
            '{"address":"6886e372363ba8718993ebf49f1a7324f9b273bd","id":"62d90545-71b7-4ed3-9338-6cb2b587f475","version":3,"Crypto":{"cipher":"aes-128-ctr","cipherparams":{"iv":"b5f0c3d1d24cd9cea36528de0605bad3"},"ciphertext":"887a2a44586a339a347379f3107580cb8c2709d832b1628d499a0cb4ad2223ef","kdf":"scrypt","kdfparams":{"salt":"c3d6770af197cecc644086460bf1f9ddb504cd993eb903e49d2ada7714b5b381","n":65536,"dklen":32,"p":1,"r":8},"mac":"b2517acaa5781bb513a0087b24cb933d7228641bb12a716fa8df50091773fb11"},"x-ethers":{"client":"ethers.js","gethFilename":"UTC--2022-07-01T06-29-13.0Z--6886e372363ba8718993ebf49f1a7324f9b273bd","mnemonicCounter":"fb5ea5b9bc34a33a0ca6deb287333f0c","mnemonicCiphertext":"35bffda94500f27dfc6245f0c3ba7cc0","path":"m/44\'/60\'/0\'/0/0","locale":"en","version":"0.1"}}',
          username: 'Test 1',
        },
      ]
    })

    render(
      <I18nextProvider i18n={i18n}>
        <CreateSubAccount />
      </I18nextProvider>
    )

    expect(screen.getByTestId('create-subaccount-btn')).toBeDisabled()
    await user.type(screen.getByTestId('username'), 'Test 2')
    await user.type(screen.getByTestId('password'), '12341234')
  })
})
