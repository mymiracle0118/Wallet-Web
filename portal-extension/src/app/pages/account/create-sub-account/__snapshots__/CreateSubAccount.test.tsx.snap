// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateSubAccount /> renders correctly Create Sub Account Page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black "
      >
        <div
          class="flex flex-row relative justify-between items-center h-16"
        >
          <button
            class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
            type="button"
          >
            <arrow-left.svg
              classname="text-xl stroke-custom-black dark:stroke-custom-white"
            />
          </button>
          <div
            class="flex-1 text-center"
          >
            <div
              class="text-custom-black dark:text-custom-white text-[18px] font-extrabold"
            >
              Create sub account
            </div>
          </div>
        </div>
        <div
          class="w-full h-full overflow-auto"
        >
          <form>
             
            <div
              class="p-4"
            >
              <div
                class="mt-6 mb-4"
              >
                <div
                  class="text-custom-grey100 dark:text-custom-white40 text-[14px] font-extrabold"
                  style="margin-bottom: 8px;"
                >
                  Onboarding.username
                </div>
                <div
                  class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
                >
                  <input
                    class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                    data-testid="username"
                    id="username"
                    name="username"
                    placeholder="Onboarding.username"
                    value=""
                  />
                  <div
                    class="pr-3.5"
                  >
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
                    >
                      <div
                        class="text-[12px] text-fotter-dark-inactive"
                      >
                        0
                        /20
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="mb-2 mt-2 text-custom-grey100 dark:text-custom-white40 text-[14px] font-extrabold"
                >
                  Account.enterPassword
                </div>
                <div
                  class="mt-4 mb-8"
                >
                  <div
                    class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
                  >
                    <input
                      class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                      data-testid="password"
                      id="enter-password"
                      name="password"
                      placeholder="Account.enterPassword"
                      type="password"
                      value=""
                    />
                    <div
                      class="pr-3.5"
                    >
                      <button
                        aria-label="toggle password visibility"
                        class="dark:stroke-custom-white stroke-custom-black w-6"
                        data-aid="passwordShowHide"
                        tabindex="-1"
                        type="button"
                      >
                        <eye-off.svg
                          classname="w-5 h-5"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    class="mt-2 ml-1 dark:text-white text-custom-grey100 font-normal text-[12px]"
                  >
                    Min 8 characters
                  </div>
                </div>
              </div>
              <div
                class="flex gap-2 justify-between mt-4"
              >
                <button
                  class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-primary dark:border-custom-white40 dark:hover:border-custom-white hover:border-primary-alt hover:text-primary-alt"
                  type="button"
                >
                  Actions.cancel
                </button>
                <button
                  class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
                  data-testid="create-subaccount-btn"
                  disabled=""
                  type="submit"
                >
                  Actions.create
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black "
    >
      <div
        class="flex flex-row relative justify-between items-center h-16"
      >
        <button
          class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
          type="button"
        >
          <arrow-left.svg
            classname="text-xl stroke-custom-black dark:stroke-custom-white"
          />
        </button>
        <div
          class="flex-1 text-center"
        >
          <div
            class="text-custom-black dark:text-custom-white text-[18px] font-extrabold"
          >
            Create sub account
          </div>
        </div>
      </div>
      <div
        class="w-full h-full overflow-auto"
      >
        <form>
           
          <div
            class="p-4"
          >
            <div
              class="mt-6 mb-4"
            >
              <div
                class="text-custom-grey100 dark:text-custom-white40 text-[14px] font-extrabold"
                style="margin-bottom: 8px;"
              >
                Onboarding.username
              </div>
              <div
                class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
              >
                <input
                  class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                  data-testid="username"
                  id="username"
                  name="username"
                  placeholder="Onboarding.username"
                  value=""
                />
                <div
                  class="pr-3.5"
                >
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
                  >
                    <div
                      class="text-[12px] text-fotter-dark-inactive"
                    >
                      0
                      /20
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="mb-2 mt-2 text-custom-grey100 dark:text-custom-white40 text-[14px] font-extrabold"
              >
                Account.enterPassword
              </div>
              <div
                class="mt-4 mb-8"
              >
                <div
                  class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
                >
                  <input
                    class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                    data-testid="password"
                    id="enter-password"
                    name="password"
                    placeholder="Account.enterPassword"
                    type="password"
                    value=""
                  />
                  <div
                    class="pr-3.5"
                  >
                    <button
                      aria-label="toggle password visibility"
                      class="dark:stroke-custom-white stroke-custom-black w-6"
                      data-aid="passwordShowHide"
                      tabindex="-1"
                      type="button"
                    >
                      <eye-off.svg
                        classname="w-5 h-5"
                      />
                    </button>
                  </div>
                </div>
                <div
                  class="mt-2 ml-1 dark:text-white text-custom-grey100 font-normal text-[12px]"
                >
                  Min 8 characters
                </div>
              </div>
            </div>
            <div
              class="flex gap-2 justify-between mt-4"
            >
              <button
                class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-primary dark:border-custom-white40 dark:hover:border-custom-white hover:border-primary-alt hover:text-primary-alt"
                type="button"
              >
                Actions.cancel
              </button>
              <button
                class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
                data-testid="create-subaccount-btn"
                disabled=""
                type="submit"
              >
                Actions.create
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
